### PR TITLE
Keycloak backend frontend integration

### PR DESCRIPTION
**Pull request by csabika98**
BACKEND:

habito-backend/habito-app/pom.xml -> 

Added

<groupId>org.springframework.security</groupId>
<artifactId>spring-security-oauth2-jose</artifactId> 

<groupId>org.springframework.boot</groupId>
<artifactId>spring-boot-starter-oauth2-resource-server</artifactId>

<groupId>org.springframework.boot</groupId>
<artifactId>spring-boot-starter-oauth2-client</artifactId>

Created a new .properties file
classpath:motivaa.properties -> for storing keycloak data independently

habito-backend/habito-app/src/main/java/com/habito/boundary/HealthCheckController.java 
Added PreAuthorize -> Making sure you can only access to if you are logged in to Keycloak

habito-backend/habito-app/src/main/java/com/habito/config/SecurityConfig.java

Very straightforward -> created new package, config -> SecurityConfig is set to ACCEPT TOKENS FROM KEYCLOAK, and YOU CAN CONTROL WHICH ENDPOINTS CAN BE PUBLIC AND WHICH ARE NOT

FRONTEND:

habito-frontend/src/App.js

Making sure we are passing the tokens to the frontend endpoint

habito-frontend/src/index.js

Added .js extension, as there were an issue when i tried to run React with NODE 21 LTS

### CODE REVIEW

File: habito-backend/habito-app/pom.xml
Comment (csabika98) [inline habito-backend/habito-app/pom.xml:1]: Added

org.springframework.security
spring-security-oauth2-jose

org.springframework.boot
spring-boot-starter-oauth2-resource-server

org.springframework.boot
spring-boot-starter-oauth2-client

File: habito-backend/habito-app/src/main/java/com/habito/HabitoBackendApplication.java
Comment (pleszr) [inline habito-backend/habito-app/src/main/java/com/habito/HabitoBackendApplication.java:1]: just marking it as this class is also impacted by motvaa.properties discussion

File: habito-backend/habito-app/src/main/java/com/habito/config/SecurityConfig.java
Comment (pleszr) [inline habito-backend/habito-app/src/main/java/com/habito/config/SecurityConfig.java:1]: question on the package, my idea was to use entity-control-boundary for the main architecture (https://en.wikipedia.org/wiki/Entity-control-boundary & explained well in https://www.youtube.com/watch?v=oJ5xCo2vzRA)

this is something we should discuss and agree together, so i think we should spend some time on defining these rules (maybe good topic for next Wednesday), wanted to share my view (as I would move the content of the config package under control).

File: habito-frontend/package.json
Comment (pleszr) [inline habito-frontend/package.json:1]: this seems strange to me. On prod, Keycloak could not be on localhost. Isnt it strange to hardcode endpoints in the code? could be i am missing something, never dealt with react. if yes, pardon my ignorance!

**Comment by csabika98** [comment]
Hello @pleszr !

The following changes have been applied:

-> motivaa.properties removed, all the entries now its in the application-local.yaml
-> package.json, proxy entry removed, .env.template created, where we can customize the PORT

**Comment by csabika98** [comment]
Hello @pleszr !

i have added new test cases to test the backend Keycloak int.

**Comment by csabika98** [comment]
Hello, @pleszr 

Detailed Changes

**src/auth/withAuth.test.js:**
Unit tests to ensure proper functioning of the withAuth component under different initialization states.

**src/components/UserInfo.test.js:**
Unit tests to verify the UserInfo component’s behavior based on different authentication states.

**src/App.test.js:**
Comprehensive tests to validate the App component’s behavior during different stages of data fetching and error handling.

**GitHub Actions Workflow (frontend-tests.yml):**
A CI/CD workflow to automate the testing of frontend components on each commit and pull request

**Tests: Backend tests now include scenarios for both authenticated and unauthenticated requests to verify the security measures.**

File: habito-backend/.idea/misc.xml
Comment (pleszr) [inline habito-backend/.idea/misc.xml:1]: is this commited by accident? I also had issues like this, but .idea should be moved out by gitignore. how does it end up here?

**Review by pleszr — CHANGES_REQUESTED** [review]

**Comment by csabika98** [comment]
**For setup():**

I had to define it at first then, i figured out its enough if we have @AutoConfigureMockMvc, but i kept it and just commented it, initially when, we are mocking or mimicing the way we would expect the App behaves in certain scenario had to define this setup()

It can be removed, its just a reminder that no need to define manually setup() anymore.. as now.. we have @AutoConfigureMockMvc ( when i last used Spring in CodeCool i used setup() i was not aware of this)

The @AutoConfigureMockMvc annotation automatically configures and injects a MockMvc instance for us.

----------------------------------------------------------------------------------------------------------------------

**For: i think this is redundant due to .andExpect(status().isUnauthorized()) ?**



I see, i made a mistake here, well, not a mistake but its not following the clean code principal but we dont need
assertEquals(401, status); because andExpect already performs this check.

In a test, using andExpect(status().isUnauthorized()) is effectively an assertion.

so
andExpect(status().isUnauthorized()) == assertEquals(401, status);

@test
void shouldDenyUnauthenticatedUserAccessToAuthTest() throws Exception {
mockMvc.perform(get("/apis/authTest"))
.andExpect(status().isUnauthorized());
}

----------------------------------------------------------------------------------------------------------------------

**For: 

I dont think this is coherent.

if we are expecting a response should we not assert it?
the printing is not in sync with the test itself**


I understand your concern.

Because the andExpect methods are assertions,
The test includes assertions for the status (isOk()) and content type (application/json)

Therefore no need to use assertEquals..

Additionally, printing expected vs. actual values can be useful for debugging, my mistake i was not sure about the expectation of this endpoint thats why i wrote {} as expected response.. in here we are comparing responses.. what would be the expectation and what is being called by the mockMVC

File: habito-backend/habito-app/src/test/java/com/habito/boundary/HealthCheckControllerTest.java [hunk 1]
```diff
@@ -1,41 +1,127 @@
...
+@Import({SecurityConfigTest.class, SecurityConfig.class})
 public class HealthCheckControllerTest {
+
     @Autowired
     private MockMvc mockMvc;
 
     @MockBean
     private HabitoService habitoService;
 
+    @BeforeEach
```
[the remark this answers is not in the document]
Comment (pleszr) [reply habito-backend/habito-app/src/test/java/com/habito/boundary/HealthCheckControllerTest.java:38]: i think best is to remove it. we will add it when needed. maybe we can add later to the documentation of the tests

File: habito-backend/habito-app/src/test/java/com/habito/boundary/HealthCheckControllerTest.java [hunk 2]
```diff
@@ -1,41 +1,127 @@
...
+        String responseJson = mockMvc.perform(get("/apis/healthcheck"))
                 .andExpect(status().isOk())
+                .andExpect(content().contentType(MediaType.APPLICATION_JSON))
                 .andReturn()
                 .getResponse()
                 .getContentAsString();
+
         String healthStatusFromJson = JsonPath.read(responseJson, "$.healthStatus");
-        assertEquals("ok", healthStatusFromJson, "Healthcheck endpoint should give ok, but it didnt");
+        System.out.println("Expected: ok, Actual: " + healthStatusFromJson);
```
[the remark this answers is not in the document]
Comment (pleszr) [reply habito-backend/habito-app/src/test/java/com/habito/boundary/HealthCheckControllerTest.java:54]: I think during development it makes sense. I would just not commit it.

I think if you want to do a custom logging, you can use log4j + trace. that way if you ever want to see them, you an make the change to trace and run the tests.

**Pull request merged**